backup: panic in url escape

I put the newline in there by accident but it panic'd the server.

```
backup database d to 'azure://host/backup?AZURE_ACCOUNT_KEY=password&AZURE_ACCOUNT_NAME=blah
';
```

```
ERROR: a panic has occurred!
If no details are printed below, check the log file for details.
panic: parse https://blah%0A.blob.core.windows.net: invalid URL escape "%0A" [recovered]
        panic: parse https://blah%0A.blob.core.windows.net: invalid URL escape "%0A" [recovered]
        panic: parse https://blah%0A.blob.core.windows.net: invalid URL escape "%0A"

goroutine 2686266 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4202ec640)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x94
panic(0x1b360e0, 0xc421f89920)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send.func1(0xc425b85538, 0xc425b855b0, 0x14b089c9327a5a38, 0x0, 0xc425b855a8, 0xc4204880
00)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2504 +0x224
panic(0x1b360e0, 0xc421f89920)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/vendor/github.com/Azure/azure-sdk-for-go/storage.Client.getEndpoint(0x0, 0xc420480d65, 0x15, 0xc420a15e00, 0x
40, 0x42, 0x1, 0x1cc622f, 0x10, 0x1cbe2dc, ...)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/Azure/azure-sdk-for-go/storage/client.go:232 +0x440
github.com/cockroachdb/cockroach/vendor/github.com/Azure/azure-sdk-for-go/storage.BlobStorageClient.CreateBlockBlobFromReader(0x0, 0xc420480d6
5, 0x15, 0xc420a15e00, 0x40, 0x42, 0x1, 0x1cc622f, 0x10, 0x1cbe2dc, ...)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/Azure/azure-sdk-for-go/storage/blob.go:1040 +0x21d
github.com/cockroachdb/cockroach/vendor/github.com/Azure/azure-sdk-for-go/storage.BlobStorageClient.CreateBlockBlob(0x0, 0xc420480d65, 0x15, 0
xc420a15e00, 0x40, 0x42, 0x1, 0x1cc622f, 0x10, 0x1cbe2dc, ...)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/Azure/azure-sdk-for-go/storage/blob.go:1026 +0xcd
github.com/cockroachdb/cockroach/pkg/ccl/storageccl.(*azureStorage).WriteFile.func1(0xc425b82d40, 0xc42c394d01)
        /go/src/github.com/cockroachdb/cockroach/pkg/ccl/storageccl/export_storage.go:413 +0xc1
github.com/cockroachdb/cockroach/pkg/util/retry.WithMaxAttempts(0x7fac57f612b0, 0xc42c394d20, 0x2faf080, 0x12a05f200, 0x4000000000000000, 0x2,
 0x0, 0x0, 0x3, 0xc425b82e28, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:151 +0xe9
github.com/cockroachdb/cockroach/pkg/ccl/storageccl.(*azureStorage).WriteFile(0xc423ebe990, 0x7fac57f612b0, 0xc42c394d20, 0xc424d4a1c0, 0x16, 
0x292ea20, 0xc4298a0040, 0xc42feb3b20, 0xc4298a0160)
        /go/src/github.com/cockroachdb/cockroach/pkg/ccl/storageccl/export_storage.go:414 +0x1c3
github.com/cockroachdb/cockroach/pkg/ccl/storageccl.(*tmpWriter).Finish(0xc42c394ea0, 0x7fac57f612b0, 0xc42c394d20, 0x40, 0x40)
        /go/src/github.com/cockroachdb/cockroach/pkg/ccl/storageccl/export_storage.go:603 +0x6e
github.com/cockroachdb/cockroach/pkg/ccl/storageccl.evalExport(0x7fac57f612b0, 0xc42c394d20, 0x7fac57f71aa8, 0xc4204942c0, 0xc4207f0700, 0xc42
8b5e6c0, 0x14b089c92a6cadc2, 0x0, 0x100000001, 0x2, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/ccl/storageccl/export.go:141 +0xe14
github.com/cockroachdb/cockroach/pkg/storage.executeCmd(0x7fac57f612b0, 0xc43405ee70, 0x0, 0x0, 0x0, 0x7fac57f71aa8, 0xc4204942c0, 0xc4207f070
0, 0xc428b5e6c0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_command.go:187 +0x301
github.com/cockroachdb/cockroach/pkg/storage.executeBatch(0x7fac57f612b0, 0xc43405ee70, 0x0, 0x0, 0x7fac57f71aa8, 0xc4204942c0, 0xc4207f0700, 
0xc428b5e6c0, 0x0, 0x14b089c92a6cadc2, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4372 +0x43c
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).addReadOnlyCmd(0xc4207f0700, 0x7fac57f612b0, 0xc43405ee70, 0x14b089c92a6cadc2, 0x0, 0x
100000001, 0x2, 0x367a, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2059 +0x2c2
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).Send(0xc4207f0700, 0x7fac57f612b0, 0xc43405ee70, 0x14b089c92a6cadc2, 0x0, 0x100000001,
 0x2, 0x367a, 0x0, 0x0, ...)
```

--- a/pkg/ccl/sqlccl/backup_test.go
+++ b/pkg/ccl/sqlccl/backup_test.go
@@ -1372,3 +1372,27 @@ func TestBackupRestorePermissions(t *testing.T) {
 		}
 	})
 }
+
+func TestBackupAzureAccountName(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	const numAccounts = 1
+	_, _, _, sqlDB, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts)
+	defer cleanupFn()
+
+	values := url.Values{}
+	values.Set("AZURE_ACCOUNT_KEY", "password")
+	values.Set("AZURE_ACCOUNT_NAME", "\n")
+
+	url := &url.URL{
+		Scheme:   "azure",
+		Host:     "host",
+		Path:     "/backup",
+		RawQuery: values.Encode(),
+	}
+
+	// Verify newlines in the account name cause an error.
+	if _, err := sqlDB.DB.Exec(`backup database d to $1`, url.String()); !testutils.IsError(err, "azure account name invalid") {
+		t.Fatalf("unexpected error, got %v", err)
+	}
+}

--- a/pkg/ccl/storageccl/export_storage.go
+++ b/pkg/ccl/storageccl/export_storage.go
@@ -17,6 +17,7 @@ import (
 	"net/url"
 	"os"
 	"path/filepath"
+	"regexp"
 	"strings"
 
 	gcs "cloud.google.com/go/storage"
@@ -375,9 +376,16 @@ type azureStorage struct {
 
 var _ ExportStorage = &azureStorage{}
 
+var azureAccountNameRegex = regexp.MustCompile(`^[a-z0-9]{3,24}$`)
+
 func makeAzureStorage(conf *roachpb.ExportStorage_Azure) (ExportStorage, error) {
 	if conf == nil {
 		return nil, errors.Errorf("azure upload requested but info missing")
+	}
+	// TODO(mjibson): Remove when
+	// https://github.com/Azure/azure-sdk-for-go/issues/584 is fixed.
+	if !azureAccountNameRegex.MatchString(conf.AccountName) {
+		return nil, errors.New("azure account name invalid")
 	}
 	client, err := azr.NewBasicClient(conf.AccountName, conf.AccountKey)
 	if err != nil {

